eZ/Publish/Core/Repository/ContentService.php

```diff
@@ -781,16 +781,22 @@ private function cloneField(Field $field, array $overrides = [])
      * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
      *
      * @param \eZ\Publish\API\Repository\Values\Content\LocationCreateStruct[] $locationCreateStructs
+     * @param \eZ\Publish\API\Repository\Values\ContentType\ContentType|null
      *
      * @return \eZ\Publish\SPI\Persistence\Content\Location\CreateStruct[]
      */
-    protected function buildSPILocationCreateStructs(array $locationCreateStructs)
+    protected function buildSPILocationCreateStructs(array $locationCreateStructs, ContentType $contentType = null)
     {
         $spiLocationCreateStructs = array();
         $parentLocationIdSet = array();
         $mainLocation = true;
 
         foreach ($locationCreateStructs as $locationCreateStruct) {
+            if ($contentType) {
+                $locationCreateStruct->sortField = $contentType->defaultSortField;
```

Comment:
This will break any value API user has set on these, and there is no sane way to detect this case.

So this needs to be on AP, kind of like what was done in https://github.com/ezsystems/ezpublish-kernel/commit/843de1a69c50943ec996cade94d5ff1614907369#diff-4321ba086690f6f89abef111ea256618

What we can do is add it as optional parameter like you did for SPI here, and log deprecation warning when it's not provided so we can make it required in 8.x.

Reply:
@alongosz I don't think this is possible as `LocationCreateStruct` has `sortField` and `sortOrder` set by default to `SORT_FIELD_NAME` and `SORT_ORDER_ASC` (see [here](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/Values/Content/LocationCreateStruct.php#L51)), so `sortField` and `sortOrder` will never be `null` as long as user won't set it like that intentionally (or not ;) ). So there is no proper way to detect if those values in `LocationCreateStruct` are set by the user or just defaults.

Reply:
Right now I'm leaning toward changes in `LocationService:newLocationCreateStruct` to set it like this:
`public function newLocationCreateStruct($parentLocationId, ContentType $contentType = null)` and as @andrerom said add entry to log deprecation warning when it's `$contentType` is not provided. As this is optional parameter I don't think it's breaking interface.

Reply:
👍